diff --git a/lib/logstash/outputs/elasticsearch/protocol.rb b/lib/logstash/outputs/elasticsearch/protocol.rb
index 3b012d1..11d17b3 100644
--- a/lib/logstash/outputs/elasticsearch/protocol.rb
+++ b/lib/logstash/outputs/elasticsearch/protocol.rb
@@ -98,6 +98,17 @@ module LogStash::Outputs::Elasticsearch
 
       def bulk(actions)
         bulk_response = @client.bulk(:body => actions.collect do |action, args, source|
+          #(sn0wtiger) update & upsert requests with fallback to index request if no id or type are given
+          if (action == 'update' || action == 'upsert') && (!args[:_id] || args[:_id].nil?)
+            action = 'index'
+          else
+            if source && action == 'update'
+              source = { 'doc' => source }
+            elsif source && action == 'upsert'
+              source = { 'doc' => source, 'doc_as_upsert' => true }
+              action = 'update'
+            end
+          end
           if source
             next [ { action => args }, source ]
           else
